swift/extractor/remapping/SwiftOpenInterception.macOS.cpp

```diff
@@ -0,0 +1,79 @@
+#include "swift/extractor/remapping/SwiftOpenInterception.h"
+#include <fishhook.h>
+#include <llvm/Support/raw_ostream.h>
+#include <llvm/Support/FileSystem.h>
+#include <llvm/Support/Path.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+namespace codeql {
+
+static std::string scratchDir;
```

Comment:
I _think_ there might be a use-after-destructor problem with this, in case `open` is called after `exit` by some other destructor of an object with static storage duration. Even though this is not happening now, we cannot really guarantee this won't happen. For example, what if the swift library initializes some non-POD static global that writes down diagnostics at the end of the program?

We could either:
* undo the interception more or less at the time of calling `remapArtifacts`, putting `original_open` back to `open`
* use a `static const char*` for the `scratchDir`, making `codeql_open` auto fallback on `original_open` if `scratchDir == nullptr`, and set it back to `nullptr` in `remapArtifacts`. The `SwiftExtractorConfiguration` object owning the scratch dir string will be alive between calls to `initInterception` and `remapArtifacts`

I'm always wary of using non-const globals (which is probably unavoidable here because of having to interface with C via free functions), but the wariness doubles when using non-POD globals (const or non const) because of the subtle bugs that can happen during end of program destructor calls. So if there is a way to avoid that I would rather take it, and go for the second option above.

Reply:
It's a very good point and it would've been a nightmare to catch/debug afterwards!

However, `SwiftExtractorConfiguration` doesn't really own the scratch dir as we changed `tempArtifactDir` to be a function returning a new string every time it's called 😅. We can easily workaround this limitation, but IMO having a  `char *` is way too fragile.

I'll see what's the best way to rebind the functions back.

Reply:
Ah, good point on the const char lifetime!

Reply:
If you go for a RAII class you could use the lifetime of that class to make sure the raw pointer is alive while required.

Reply:
I played a bit with the options and the best suggestion I have is to add a boolean flag, though as I'm writing I'm wondering if it should be an atomic boolean? 🤔

Reply:
Good question, does llvm or the swift frontend fire up threads? Even if they do, are they still not joined when we execute our code?

Another option could be maybe to make `scratchDir` indestructible. Declare it as
```
static const std::string* scratchDir = nullptr;
```
and initialize it with
```
scratchDir = new std::string(dir);
```
and just fuggedaboutit... but I'm ok with this solution. We will probably need to come back to this code if we generalize the interception to Linux.

Reply:
Yeah, all the available options are more like "which footgun shall we pick" 😄